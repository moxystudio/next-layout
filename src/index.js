export { default as LayoutTree } from './LayoutTree';
export { default as withLayout } from './with-layout';
