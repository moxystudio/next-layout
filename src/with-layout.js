const layoutSymbol = Symbol('layout');

const withLayout = (mapToLayout) => (Component) => {
    const value = typeof mapToLayout === 'function' ? mapToLayout : () => mapToLayout;

    Object.defineProperty(Component, layoutSymbol, { value });

    return Component;
};

export const getLayoutFromPage = (Component, pageProps, defaultLayout) => {
    const layout = Component[layoutSymbol]?.(pageProps) ?? defaultLayout;

    return {
        Layout: layout?.type,
        layoutProps: layout?.props,
    };
};

export const shouldInjectSetLayoutProp = (Component) => !!Component[layoutSymbol];

export default withLayout;
