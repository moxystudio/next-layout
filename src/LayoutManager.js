import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getLayoutFromPage, shouldInjectSetLayoutProp } from './with-layout';

const DummyLayout = ({ children }) => children;

export default class LayoutManager extends Component {
    static propTypes = {
        Component: PropTypes.elementType.isRequired,
        pageProps: PropTypes.object,
        defaultLayout: PropTypes.element,
        children: PropTypes.func,
    };

    static defaultProps = {
        defaultLayout: <DummyLayout />,
        children: ({ Layout, layoutProps, Component, pageProps }) => (
            <Layout { ...layoutProps }>
                <Component { ...pageProps } />
            </Layout>
        ),
    };

    static getDerivedStateFromProps(props, state) {
        const { Component, pageProps, defaultLayout } = props;
        const { layoutCounter = 0, pageCounter = 0 } = state;

        const { Layout, layoutProps } = getLayoutFromPage(Component, pageProps, defaultLayout);

        const didLayoutChanged = Layout !== state.Layout;
        const didPageChanged = props.Component !== state.Component;

        return {
            Layout,
            layoutProps: didPageChanged ? layoutProps : state.layoutProps,
            layoutCounter: didLayoutChanged ? layoutCounter + 1 : layoutCounter,
            pageCounter: didPageChanged ? pageCounter + 1 : pageCounter,
            Component: props.Component,
        };
    }

    state = {};

    render() {
        const { Component, pageProps, children } = this.props;
        const { Layout, layoutProps, layoutCounter, pageCounter } = this.state;

        const layoutKey = `layout-${layoutCounter.toString()}`;
        const pageKey = `page-${pageCounter.toString()}`;
        const finalPageProps = shouldInjectSetLayoutProp(Component) ?
            { setLayoutProps: this.setLayoutProps, ...pageProps } : pageProps;

        return children({
            Layout,
            layoutProps,
            layoutKey,
            Component,
            pageProps: finalPageProps,
            pageKey,
        });
    }

    setLayoutProps = (layoutProps) => {
        if (typeof layoutProps === 'function') {
            layoutProps = layoutProps(this.state.layoutProps);
        }

        this.setState((state) => ({
            layoutProps: { ...state.layoutProps, ...layoutProps },
        }));
    };
}
