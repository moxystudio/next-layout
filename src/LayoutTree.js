import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import LayoutContext from './util/context';
import createFullTree from './util/full-tree';
import { getInitialLayoutTree } from './with-layout';

const LayoutProvider = LayoutContext.Provider;

export default class LayoutTree extends PureComponent {
    static propTypes = {
        Component: PropTypes.elementType.isRequired,
        pageProps: PropTypes.object,
        defaultLayout: PropTypes.element,
        children: PropTypes.func,
    };

    static defaultProps = {
        children: (rootNode) => rootNode,
    };

    static getDerivedStateFromProps(props, state) {
        const { Component, pageProps } = props;

        const didPageChange = props.Component !== state.Component;
        const layoutTree = didPageChange ? getInitialLayoutTree(Component, pageProps) : state.layoutTree;

        return {
            Component,
            layoutTree,
        };
    }

    state = {};

    providerValue = {
        updateLayoutTree: (layoutTree) => this.setState({ layoutTree }),
    };

    render() {
        const { defaultLayout, Component, pageProps, children: render } = this.props;
        const { layoutTree } = this.state;

        const page = <Component { ...pageProps } />;
        const fullTree = createFullTree(layoutTree ?? defaultLayout, page);

        return (
            <LayoutProvider value={ this.providerValue }>
                { render(fullTree) }
            </LayoutProvider>
        );
    }
}
