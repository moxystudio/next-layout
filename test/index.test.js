import React, { useEffect } from 'react';
import { mount } from 'enzyme';
import { withLayout, LayoutTree } from '../src';

const PrimaryLayout = ({ children }) => <main>{ children }</main>;
const Home = () => <h1>Home</h1>;

afterEach(() => {
    console.error.mockRestore?.();
});

it('should render no layout', () => {
    const wrapper = mount(
        <LayoutTree
            Component={ Home }
            pageProps={ { foo: 'bar' } } />,
    );

    expect(wrapper).toMatchSnapshot();
});

it('should render default layout', () => {
    const wrapper = mount(
        <LayoutTree
            Component={ Home }
            pageProps={ { foo: 'bar' } }
            defaultLayout={ <PrimaryLayout /> } />,
    );

    expect(wrapper).toMatchSnapshot();
});

it('should render a one level deep layout tree correctly', () => {
    const EnhancedHome = withLayout(<PrimaryLayout />)(Home);

    const wrapper = mount(
        <LayoutTree
            Component={ EnhancedHome }
            pageProps={ { foo: 'bar' } } />,
    );

    expect(wrapper).toMatchSnapshot();
});

it('should render a two level deep layout tree correctly', () => {
    const AccountLayout = ({ children }) => <div>{ children }</div>;

    const layout = (
        <PrimaryLayout>
            <AccountLayout />
        </PrimaryLayout>
    );

    const EnhancedAccountInfo = withLayout(layout)(Home);

    const wrapper = mount(
        <LayoutTree
            Component={ EnhancedAccountInfo }
            pageProps={ { foo: 'bar' } } />,
    );

    expect(wrapper).toMatchSnapshot();
});

it('should call layout\'s and page\'s render just once', () => {
    const PrimaryLayoutMock = jest.fn(PrimaryLayout);
    const HomeMock = jest.fn(Home);

    const EnhancedHomeMock = withLayout(<PrimaryLayoutMock />)(HomeMock);

    mount(<LayoutTree Component={ EnhancedHomeMock } />);

    expect(PrimaryLayoutMock).toHaveBeenCalledTimes(1);
    expect(HomeMock).toHaveBeenCalledTimes(1);
});

it('should render a layout tree correctly based the initial layout state', () => {
    const mapLayoutStateToLayoutTree = jest.fn(({ variant }) => <PrimaryLayout variant={ variant } />);

    const EnhancedHome = withLayout(mapLayoutStateToLayoutTree, { variant: 'light' })(Home);

    const wrapper = mount(<LayoutTree Component={ EnhancedHome } />);

    expect(wrapper).toMatchSnapshot();
    expect(mapLayoutStateToLayoutTree).toHaveBeenCalledTimes(1);
    expect(mapLayoutStateToLayoutTree).toHaveBeenLastCalledWith({ variant: 'light' });
});

it('should render a layout tree correctly based the initial layout state (function)', () => {
    const mapLayoutStateToLayoutTree = jest.fn(({ variant }) => <PrimaryLayout variant={ variant } />);
    const mapPropsToInitialLayoutState = jest.fn(({ light }) => ({ variant: light ? 'light' : 'dark' }));

    const EnhancedHome = withLayout(mapLayoutStateToLayoutTree, mapPropsToInitialLayoutState)(Home);

    const wrapper = mount(
        <LayoutTree
            Component={ EnhancedHome }
            pageProps={ { light: true } } />,
    );

    expect(wrapper).toMatchSnapshot();
    expect(mapLayoutStateToLayoutTree).toHaveBeenCalledTimes(1);
    expect(mapLayoutStateToLayoutTree).toHaveBeenLastCalledWith({ variant: 'light' });
    expect(mapPropsToInitialLayoutState).toHaveBeenCalledTimes(2);
    expect(mapPropsToInitialLayoutState).toHaveBeenLastCalledWith({ light: true });
});

it('should update the layout tree correctly if setLayoutState is called', async () => {
    const Home = ({ setLayoutState }) => {
        useEffect(() => {
            setLayoutState({ variant: 'dark' });
        }, [setLayoutState]);

        return <h1>Home</h1>;
    };

    const PrimaryLayoutMock = jest.fn(PrimaryLayout);
    const HomeMock = jest.fn(Home);

    const mapLayoutStateToLayoutTree = jest.fn(({ variant }) => <PrimaryLayoutMock variant={ variant } />);

    const EnhancedHomeMock = withLayout(mapLayoutStateToLayoutTree, { variant: 'light' })(HomeMock);

    const wrapper = mount(<LayoutTree Component={ EnhancedHomeMock } />);

    await new Promise((resolve) => setTimeout(resolve, 50));

    expect(wrapper).toMatchSnapshot();
    expect(PrimaryLayoutMock).toHaveBeenCalledTimes(2);
    expect(HomeMock).toHaveBeenCalledTimes(2);
});

it('should update the layout tree correctly if setLayoutState is called (function)', async () => {
    expect.assertions(4);

    const Home = ({ setLayoutState }) => {
        useEffect(() => {
            setLayoutState((layoutState) => {
                expect(layoutState).toEqual({ variant: 'light' });

                return { variant: 'dark' };
            });
        }, [setLayoutState]);

        return <h1>Home</h1>;
    };

    const PrimaryLayoutMock = jest.fn(PrimaryLayout);
    const HomeMock = jest.fn(Home);

    const mapLayoutStateToLayoutTree = jest.fn(({ variant }) => <PrimaryLayoutMock variant={ variant } />);

    const EnhancedHomeMock = withLayout(mapLayoutStateToLayoutTree, { variant: 'light' })(HomeMock);

    const wrapper = mount(<LayoutTree Component={ EnhancedHomeMock } />);

    await new Promise((resolve) => setTimeout(resolve, 50));

    expect(wrapper).toMatchSnapshot();
    expect(PrimaryLayoutMock).toHaveBeenCalledTimes(2);
    expect(HomeMock).toHaveBeenCalledTimes(2);
});

it('should not inject setLayoutState to pages if it is not needed', () => {
    const HomeMock = jest.fn(Home);
    const EnhancedHome = withLayout(<PrimaryLayout />)(HomeMock);

    mount(<LayoutTree Component={ EnhancedHome } />);

    expect(HomeMock).toHaveBeenCalledTimes(1);
    expect(HomeMock).toHaveBeenCalledWith({}, {});
});

it('should fail if layout tree is not unary', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const NewsletterForm = () => <form />;
    const AccountLayout = ({ children }) => <div>{ children }</div>;

    const layout = (
        <PrimaryLayout>
            <AccountLayout />
            <NewsletterForm />
        </PrimaryLayout>
    );

    const EnhancedHome = withLayout(layout)(Home);

    expect(() => mount(<LayoutTree Component={ EnhancedHome } />)).toThrow(/unary trees/i);
});

it('should allow passing a custom render prop to LayoutTree', () => {
    const EnhancedHome = withLayout(<PrimaryLayout />)(Home);

    const render = jest.fn((tree) => tree);

    const wrapper = mount(
        <LayoutTree
            Component={ EnhancedHome }
            pageProps={ { foo: 'bar' } }>
            { render }
        </LayoutTree>,
    );

    expect(wrapper).toMatchSnapshot();
    expect(render).toHaveBeenCalledTimes(1);
    expect(render.mock.results[0].value).toMatchSnapshot();
});
