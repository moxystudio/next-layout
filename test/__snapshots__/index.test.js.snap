// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow passing a custom render prop to LayoutTree 1`] = `
<LayoutTree
  Component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "WithLayout(Home)",
      "render": [Function],
    }
  }
  pageProps={
    Object {
      "foo": "bar",
    }
  }
>
  <PrimaryLayout>
    <main>
      <WithLayout(Home)
        foo="bar"
      >
        <Home
          foo="bar"
        >
          <h1>
            Home
          </h1>
        </Home>
      </WithLayout(Home)>
    </main>
  </PrimaryLayout>
</LayoutTree>
`;

exports[`should allow passing a custom render prop to LayoutTree 2`] = `
<PrimaryLayout>
  <WithLayout(Home)
    foo="bar"
  />
</PrimaryLayout>
`;

exports[`should render a layout tree correctly based the initial layout state (function) 1`] = `
<LayoutTree
  Component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "WithLayout(Home)",
      "render": [Function],
    }
  }
  pageProps={
    Object {
      "light": true,
    }
  }
>
  <PrimaryLayout
    variant="light"
  >
    <main>
      <WithLayout(Home)
        light={true}
      >
        <Home
          light={true}
          setLayoutState={[Function]}
        >
          <h1>
            Home
          </h1>
        </Home>
      </WithLayout(Home)>
    </main>
  </PrimaryLayout>
</LayoutTree>
`;

exports[`should render a layout tree correctly based the initial layout state 1`] = `
<LayoutTree
  Component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "WithLayout(Home)",
      "render": [Function],
    }
  }
>
  <PrimaryLayout
    variant="light"
  >
    <main>
      <WithLayout(Home)>
        <Home
          setLayoutState={[Function]}
        >
          <h1>
            Home
          </h1>
        </Home>
      </WithLayout(Home)>
    </main>
  </PrimaryLayout>
</LayoutTree>
`;

exports[`should render a one level deep layout tree correctly 1`] = `
<LayoutTree
  Component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "WithLayout(Home)",
      "render": [Function],
    }
  }
  pageProps={
    Object {
      "foo": "bar",
    }
  }
>
  <PrimaryLayout>
    <main>
      <WithLayout(Home)
        foo="bar"
      >
        <Home
          foo="bar"
        >
          <h1>
            Home
          </h1>
        </Home>
      </WithLayout(Home)>
    </main>
  </PrimaryLayout>
</LayoutTree>
`;

exports[`should render a two level deep layout tree correctly 1`] = `
<LayoutTree
  Component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "WithLayout(Home)",
      "render": [Function],
    }
  }
  pageProps={
    Object {
      "foo": "bar",
    }
  }
>
  <PrimaryLayout>
    <main>
      <AccountLayout>
        <div>
          <WithLayout(Home)
            foo="bar"
          >
            <Home
              foo="bar"
            >
              <h1>
                Home
              </h1>
            </Home>
          </WithLayout(Home)>
        </div>
      </AccountLayout>
    </main>
  </PrimaryLayout>
</LayoutTree>
`;

exports[`should render default layout 1`] = `
<LayoutTree
  Component={[Function]}
  defaultLayout={<PrimaryLayout />}
  pageProps={
    Object {
      "foo": "bar",
    }
  }
>
  <PrimaryLayout>
    <main>
      <Home
        foo="bar"
      >
        <h1>
          Home
        </h1>
      </Home>
    </main>
  </PrimaryLayout>
</LayoutTree>
`;

exports[`should render no layout 1`] = `
<LayoutTree
  Component={[Function]}
  pageProps={
    Object {
      "foo": "bar",
    }
  }
/>
`;

exports[`should update the layout tree correctly if setLayoutState is called (function) 1`] = `
<LayoutTree
  Component={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "setLayoutState": [Function],
          },
          Object {},
        ],
        Array [
          Object {
            "setLayoutState": [Function],
          },
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <h1>
            Home
          </h1>,
        },
        Object {
          "type": "return",
          "value": <h1>
            Home
          </h1>,
        },
      ],
    }
  }
>
  <mockConstructor
    variant="dark"
  >
    <main>
      <WithLayout(mockConstructor)>
        <mockConstructor
          setLayoutState={[Function]}
        >
          <h1>
            Home
          </h1>
        </mockConstructor>
      </WithLayout(mockConstructor)>
    </main>
  </mockConstructor>
</LayoutTree>
`;

exports[`should update the layout tree correctly if setLayoutState is called 1`] = `
<LayoutTree
  Component={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "setLayoutState": [Function],
          },
          Object {},
        ],
        Array [
          Object {
            "setLayoutState": [Function],
          },
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <h1>
            Home
          </h1>,
        },
        Object {
          "type": "return",
          "value": <h1>
            Home
          </h1>,
        },
      ],
    }
  }
>
  <mockConstructor
    variant="light"
  >
    <main>
      <WithLayout(mockConstructor)>
        <mockConstructor
          setLayoutState={[Function]}
        >
          <h1>
            Home
          </h1>
        </mockConstructor>
      </WithLayout(mockConstructor)>
    </main>
  </mockConstructor>
</LayoutTree>
`;
